sys/include/sema.h

```diff
@@ -33,21 +35,37 @@ extern "C" {
 /**
  * @brief   Creates semaphore statically.
  *
- * @param[in] value Initial value for the semaphore.
+ * The value for the statically created semaphore must be 0
  *
  * @return  Statically initialized semaphore.
  */
-#define SEMA_CREATE(value)      { (value), PRIORITY_QUEUE_INIT }
+#define SEMA_CREATE()           { (0), SEMA_OK, MUTEX_INIT() }
+
+/**
+ * @brief A Semaphore states.
+ */
+typedef enum {
+    SEMA_OK = 0,
+    SEMA_DESTROY,
+} sema_state_t;
 
 /**
  * @brief A Semaphore.
  */
 typedef struct {
-    volatile unsigned int value;    /**< value of the semaphore */
-    priority_queue_t queue;         /**< list of threads waiting for the semaphore */
+    atomic_int_t value;             /**< value of the semaphore */
+    sema_state_t state;             /**< state of the semaphore */
+    mutex_t mutex;                  /**< mutex of the semaphore */
```

Comment:
Before it was `unsigned int` + `priority_queue_node_t`, assuming a 32-bit processor this is `4 + 12 = 16` byte. Now it is `unsigned int` + `enum` + `mutex_t` which is `4+4+4 = 12` bytes. This is 4 byte less. One could remove the `enum` and use `signed int` for the value (negative meaning destroy) and could save 4 bytes more (the posix implementation uses a signed int).

I liked the point to remove the flags dependency because, adding the flags adds 2 bytes to every thread stack even if the semaphore is only used by some of the threads.